src/login/msal/MsalAuthProvider.ts

```diff
@@ -0,0 +1,122 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Environment } from "@azure/ms-rest-azure-env";
+import { AzureIdentityCredentialAdapter } from '@azure/ms-rest-js';
+import { AccountInfo, AuthenticationResult, Configuration, LogLevel, PublicClientApplication, TokenCache } from "@azure/msal-node";
+import { MemoryCache } from "adal-node";
+import { DeviceTokenCredentials } from "ms-rest-azure";
+import { ExtensionContext } from "vscode";
+import { AzureSession } from "../../azure-account.api";
+import { clientId, msalScopes } from "../../constants";
+import { AzureLoginError } from "../../errors";
+import { localize } from "../../utils/localize";
+import { ProxyTokenCache } from "../adal/tokens";
+import { AbstractCredentials2, AbstractLoginResult, AuthProviderBase, isAdalLoginResult, loginResultTypeError } from "../AuthProviderBase";
+import { cachePlugin } from "./cachePlugin";
+import { PublicClientCredential } from "./PublicClientCredential";
+
+export class MsalAuthProvider extends AuthProviderBase {
+	private publicClientApp: PublicClientApplication;
+
+	// For compatibility with `DeviceTokenCredentials`
+	private dummyTokenCache: ProxyTokenCache = new ProxyTokenCache(new MemoryCache());
+
+	constructor(context: ExtensionContext, enableVerboseLogs: boolean) {
+		super(context);
+		const msalConfiguration: Configuration = {
+			auth: { clientId },
+			cache: { cachePlugin },
+			system: {
+				loggerOptions: {
+					loggerCallback: (_level: LogLevel, message: string, _containsPii: boolean) => {
+						this.outputChannel.appendLine(message);	
+					},
+					piiLoggingEnabled: false,
+					logLevel: enableVerboseLogs ? LogLevel.Verbose : LogLevel.Error
+				}
+			}
+		};
+		this.publicClientApp = new PublicClientApplication(msalConfiguration);
+	}
+
+	public async loginWithoutLocalServer(_clientId: string, _environment: Environment, _isAdfs: boolean, _tenantId: string): Promise<AbstractLoginResult> {
+		throw new Error('"Login Without Local Server" not implemented for MSAL.');
```

Comment:
To avoid bloating this PR too much I'm waiting to implement `loginWithoutLocalServer` and `loginWithDeviceCode` for MSAL